csrc/executor.cpp

```diff
@@ -963,56 +963,60 @@ at::Tensor allocateOutput(
     return ee.evaluate(out_tv).as<at::Tensor>();
   }
 
-  if (alias_info.type == AllocationType::NoAlias) {
-    auto alloc_tensor = at::native::empty_strided_cuda(
-        out_info.sizes,
-        out_info.strides,
-        out_info.type,
-        c10::nullopt,
-        device,
-        c10::nullopt);
-    if (shouldFillAllocationWithNan()) {
-      fillTensorWithNan(alloc_tensor);
+  std::optional<at::Tensor> aliased_io_tensor = std::nullopt;
+  Val* aliased_io = alias_info.aliased_io;
+  if (aliased_io != nullptr) {
+    NVF_ERROR(
+        aliased_io->isFusionInput() || aliased_io->isFusionOutput(),
+        aliased_io->toInlineString(),
+        " is expected to be a fusion input/output. `ee.evaluate` ",
+        "an intermediate tensor may involve GPU computation to materialize it ",
+        "to global memory.");
+    const PolymorphicValue& aliased_io_val = ee.evaluate(aliased_io);
+    NVF_ERROR(
+        aliased_io_val.is<at::Tensor>(),
+        "Alias io only supports tensor. Found ",
+        PolymorphicValue_functions::toString(aliased_io_val));
+    aliased_io_tensor = aliased_io_val.as<at::Tensor>();
+  }
+
+  switch (alias_info.type) {
+    case AllocationType::NoAlias: {
+      auto alloc_tensor = at::native::empty_strided_cuda(
+          out_info.sizes,
+          out_info.strides,
+          out_info.type,
+          c10::nullopt,
+          device,
+          c10::nullopt);
+      if (shouldFillAllocationWithNan()) {
+        fillTensorWithNan(alloc_tensor);
+      }
+      return alloc_tensor;
     }
-    return alloc_tensor;
+    case AllocationType::InplaceUpdate:
+      // Unlike for `AllocationType::Evaluate`, don't use
+      // ExpressionEvaluator to compute the output tensor. This is because
+      // the output tensor may hold different data from the input, e.g., an
+      // updated running mean.  `ExpressionEvaluator::evaluate(out_tv)`
+      // would trigger non-trivial host computation.
+      return aliased_io_tensor.value();
+    case AllocationType::Evaluate: {
+      auto out_tensor = ee.evaluate(out_tv).as<at::Tensor>();
+      if (aliased_io_tensor.has_value()) {
+        NVF_ERROR(
+            out_tensor.is_alias_of(aliased_io_tensor.value()),
+            "ExpressionEvaluator failed to evaluate ",
+            out_tv->toString(),
+            " as an alias of ",
+            aliased_io->toString());
+        inferAndValidateAllocationSizesAndStrides(out_tensor, out_tv, ee);
+      }
+      return out_tensor;
+    }
+    default:
+      NVF_ERROR(false, "Unrecognized AllocationType.");
```

Comment:
I doubt this is needed. Compilers today should be smart enough to figure out the above cases have covered all possible options.

Reply:
Added this to avoid the error: `error: control reaches end of non-void function` even though the switch-case is completely handled.